frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -58,16 +106,154 @@ impl<StorageType, T, I, L> Default for Storage<StorageType, T, I, L> {
+
+		// Effectively move a rolling window of fork-unique leaves. Once out of the window, leaves
+		// are "canonicalized" in offchain by moving them under `Pallet::node_canon_offchain_key`.
+		let leaves = NumberOfLeaves::<T, I>::get();
+		let window_size =
+			<T as frame_system::Config>::BlockHashCount::get().unique_saturated_into();
+		if leaves >= window_size {
+			// Move the rolling window towards the end of  `block_num->hash` mappings available
+			// in the runtime: we "canonicalize" the leaf at the end,
+			let to_canon_leaf = leaves.saturating_sub(window_size);
+			// and all the nodes added by that leaf.
+			let to_canon_nodes = NodesUtils::right_branch_ending_in_leaf(to_canon_leaf);
+			frame_support::log::debug!(
+				target: "runtime::mmr", "Nodes to canon for leaf {}: {:?}",
+				to_canon_leaf, to_canon_nodes
+			);
+			// For this block number there may be node entries saved from multiple forks.
+			let to_canon_block_num =
+				Pallet::<T, I>::leaf_index_to_parent_block_num(to_canon_leaf, leaves);
+			// Only entries under this hash (retrieved from state on current canon fork) are to be
+			// persisted. All other entries added by same block number will be cleared.
+			let to_canon_hash = <frame_system::Pallet<T>>::block_hash(to_canon_block_num);
+
+			Self::canonicalize_nodes_for_hash(&to_canon_nodes, to_canon_hash);
+			// Get all the forks to prune, also remove them from the offchain pruning_map.
+			PruningMap::<T, I>::remove(to_canon_block_num)
+				.map(|forks| {
+					Self::prune_nodes_for_forks(&to_canon_nodes, forks);
+				})
+				.unwrap_or_else(|| {
+					frame_support::log::debug!(
+						target: "runtime::mmr",
+						"Offchain: could not prune: no entry in pruning map for block {:?}",
+						to_canon_block_num
+					);
+				})
+		}
+	}
+
+	fn prune_nodes_for_forks(nodes: &[NodeIndex], forks: Vec<<T as frame_system::Config>::Hash>) {
+		for hash in forks {
+			for pos in nodes {
+				let key = Pallet::<T, I>::node_offchain_key(hash, *pos);
+				offchain::local_storage_clear(StorageKind::PERSISTENT, &key);
+			}
+		}
+	}
+
+	fn canonicalize_nodes_for_hash(
+		to_canon_nodes: &[NodeIndex],
+		to_canon_hash: <T as frame_system::Config>::Hash,
+	) {
+		for pos in to_canon_nodes {
+			let key = Pallet::<T, I>::node_offchain_key(to_canon_hash, *pos);
+			// Retrieve the element from Off-chain DB under fork-aware key.
+			if let Some(elem) = offchain::local_storage_get(StorageKind::PERSISTENT, &key) {
+				let canon_key = Pallet::<T, I>::node_canon_offchain_key(*pos);
+				// Add under new canon key.
+				offchain::local_storage_set(StorageKind::PERSISTENT, &canon_key, &elem);
+				frame_support::log::debug!(
+					target: "runtime::mmr",
+					"Moved elem at pos {} from key {:?} to canon key {:?}",
+					pos, key, canon_key
+				);
+			} else {
+				frame_support::log::debug!(
+					target: "runtime::mmr",
+					"Offchain: could not canonicalize elem at pos {} using key {:?}",
+					pos, key
+				);
+			}
```

Comment:
If some node cannot be found under the current canonical fork's key, it's not (re)inserted to offchain db storage under a canonical key, but given the subsequent handling in [`canonicalize_and_prune(...)`](https://github.com/paritytech/substrate/pull/11594/files#diff-0dfce96fe6bc14a20784c7bd3c9104c1e5dd1da11e7e732e06c9bec632c91be8R156-R167), it now disappears entirely from the offchain db since all of its fork contenders are pruned from the offchain db.

This strikes me as correct, but I also don't have enough experience here to be aware of subtle failure cases (aside from database corruption etc.), where a node would only be present for some forks but not others (ignoring any forks that stalled prior to the node's rightmost child-leaf being added).

So I want to confirm that this is the desired behaviour - else I'd suggest handling the node lookup failure, or logging [this](https://github.com/paritytech/substrate/pull/11594/files#diff-0dfce96fe6bc14a20784c7bd3c9104c1e5dd1da11e7e732e06c9bec632c91be8R197-R201) at the error level instead.

Reply:
That's a good point. This branch should never be hit, unless there's a programming/logic error.
Let's log it as `error` so we make it visible in case it ever happens. 👍